### PR TITLE
video_core: Fix moltenvk crash on macos 

### PR DESCRIPTION
**Pull request by vonchenplus**
1. Set vertex buffer num to 16, because mvk have when using more than 16.
2. Add missing scissor update when viewport scale offset disable.
3. Fix d24r8/s8d24 convert shader build error in moltenvk

### CODE REVIEW

**Comment by Kelebek1** [comment]
I already did this and it breaks games, I had to revert it.

**Comment by gidoly** [comment]
seems cause no issues from my end might worth ea tag and let people test it

**Comment by vonchenplus** [comment]
> I already did this and it breaks games, I had to revert it with [e681f56](https://github.com/yuzu-emu/yuzu/commit/e681f5678c3df6d2e1779cb824615517ddae46d7).
> 
> Did you test this PR with a load of games?

Which game had the problem before?

**Comment by Kelebek1** [comment]
Tested and can confirm this breaks #10802 again

**Comment by Kelebek1** [comment]
Ok so the dirty flags only track changes to the 3d registers affecting binding, they say nothing about the state of the data being pointed to by a bind region. The same vertex buffer binding may be valid between draws, but was written to by the guest and its data needs uploading regardless of rebinding.

I saw a few things though, firstly binding sizes of 0 are already excluded in UpdateVertexBuffers, and a NULL_BINDING is used instead. We can check for that null binding ID instead just to make it more clear. I think this will work more as you want and fixes the bug:

```diff
diff --git a/src/video_core/buffer_cache/buffer_cache.h b/src/video_core/buffer_cache/buffer_cache.h
index f0f450edb..15b919b83 100644
--- a/src/video_core/buffer_cache/buffer_cache.h
+++ b/src/video_core/buffer_cache/buffer_cache.h
@@ -758,9 +758,20 @@ void BufferCache<P>::BindHostVertexBuffers() {
     auto& flags = maxwell3d->dirty.flags;
     for (u32 index = 0; index < NUM_VERTEX_BUFFERS; ++index) {
         const Binding& binding = channel_state->vertex_buffers[index];
+        if (binding.buffer_id == NULL_BINDING.buffer_id) {
+            continue;
+        }
+
         Buffer& buffer = slot_buffers[binding.buffer_id];
         TouchBuffer(buffer, binding.buffer_id);
+
+        auto cpu_modified = IsRegionCpuModified(binding.cpu_addr, binding.size);
+        if (!flags[Dirty::VertexBuffer0 + index] && !cpu_modified) {
+            continue;
+        }
+
         SynchronizeBuffer(buffer, binding.cpu_addr, binding.size);
+
         if (!flags[Dirty::VertexBuffer0 + index]) {
             continue;
         }
@@ -774,8 +785,6 @@ void BufferCache<P>::BindHostVertexBuffers() {
     if (any_valid) {
         host_bindings.max_index++;
         for (u32 index = host_bindings.min_index; index < host_bindings.max_index; index++) {
-            flags[Dirty::VertexBuffer0 + index] = false;
-
             const Binding& binding = channel_state->vertex_buffers[index];
             Buffer& buffer = slot_buffers[binding.buffer_id];
```

File: src/video_core/renderer_vulkan/vk_rasterizer.cpp [hunk 1]
```diff
@@ -989,6 +989,19 @@ void RasterizerVulkan::UpdateScissorsState(Tegra::Engines::Maxwell3D::Regs& regs
...
+    if (!regs.viewport_scale_offset_enabled) {
+        const auto x = static_cast<float>(regs.surface_clip.x);
+        const auto y = static_cast<float>(regs.surface_clip.y);
+        const auto width = static_cast<float>(regs.surface_clip.width);
+        const auto height = static_cast<float>(regs.surface_clip.height);
+        VkRect2D scissor;
+        scissor.offset.x = x;
+        scissor.offset.y = y;
+        scissor.extent.width = width != 0.0f ? width : 1.0f;
+        scissor.extent.height = height != 0.0f ? height : 1.0f;
```
Comment (Kelebek1) [inline src/video_core/renderer_vulkan/vk_rasterizer.cpp:1001]: You'll need to consider rescaling here, like it is below.

**Review by liamwhite — APPROVED** [review]

**Pull request merged**